### ISSUE TITLE
pip.installed state not found

### ISSUE THREAD

**Issue by nitay** [saltstack/salt#26493]
salt master:
```
root@salt:/srv# salt-call --versions
                  Salt: 2015.5.3
                Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: 0.9.1
               Tornado: Not Installed
 Debian source package: 2015.5.3+ds-1trusty1
```

minion:
```
root@m1:~# salt-call --versions
                  Salt: 2015.5.3
                Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: Not Installed
               Tornado: 4.2.1
 Debian source package: 2015.5.3+ds-1trusty1
```

State file (snipped to minimal):
```
util-pkgs:
  pkg.installed:
    - names:
      - python-pip

pip-utils:
  pip.installed:
    - names:
      - awscli
    - require:
      - pkg: python-pip
```

For some reason salt is not able to use pip anymore..

```
KeyError: 'pip.installed'
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 227, in salt_call
    client.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/call.py", line 69, in run
    caller.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 236, in run
    ret = self.call()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 138, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/modules/state.py", line 523, in highstate
    whitelist=kwargs.get('whitelist')
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2968, in call_highstate
    return self.state.call_high(high)
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2067, in call_high
    ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1623, in call_chunks
    running = self.call_chunk(low, running, chunks)
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1769, in call_chunk
    self._mod_init(low)
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 612, in _mod_init
    self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
  File "/usr/lib/python2.7/dist-packages/salt/utils/lazy.py", line 90, in __getitem__
    raise KeyError(key)
KeyError: 'pip.installed'
```

Even though pip seems to be installed:

```
root@m1:~# python
Python 2.7.6 (default, Jun 22 2015, 17:58:13)
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pip
>>> import pip.req
>>>
```

What's going on here? This is a MAJOR blocker for me b/c it's breaking a bunch of states and making salt unusable.

**Comment by ssgward** [comment]
@nitay - Interesting. I just ran a couple of pip.installed states using v2015.5.4 tag and didn't have any problems. Wonder if something happened after 2015.5.3 to fix the problem you are seeing...?

Here's one of my states:
```
include: 
  - pip

psycopg2:
  pip.installed:
    - name: psycopg2 == 2.6
    - require:
      - sls: pip
```
What does ``pip --version`` show you?

**Comment by nitay** [comment]
master:
```
root@salt:/srv# pip --version
pip 7.1.0 from /usr/local/lib/python2.7/dist-packages (python 2.7)
```

minion:
```
root@m1:~# pip --version
pip 1.5.4 from /usr/lib/python2.7/dist-packages (python 2.7)
```

**Comment by nitay** [comment]
@ssgward do you think it's an issue with the version? I have other minions with that version that *do* work. I can't tell what makes them different...?

**Comment by ssgward** [comment]
@nitay - it maybe the version. It may have been fixed between the versions.

**Comment by nitay** [comment]
@ssgward how does that explain it working on some minions and not on others? All my machines have the same salt version.

**Comment by ssgward** [comment]
@nitay - oh, sorry, didn't read well enough to understand all minions are on the same version. I thought a few were on a newer version.

That is confusing as to why some minions would work and others not. Don't have an immediate answer for that.

**Comment by nitay** [comment]
@ssgward how do I investigate this? I'm at a loss for what to do.

**Comment by ssgward** [comment]
@nitay - I'll add labels to this issue so it will then get picked up by a dev, either internal to salt stack or a contributor in the near future.

Reaching out in forums may shed some light on this for you. Others may have seen the same thing as you are seeing.

**Comment by waynew** [comment]
I'm getting the same result on this minion:
```
                      Salt: 2015.5.3
                    Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
                    Jinja2: 2.7.2
                  M2Crypto: 0.21.1
            msgpack-python: 0.3.0
              msgpack-pure: Not Installed
                  pycrypto: 2.6.1
                   libnacl: Not Installed
                    PyYAML: 3.10
                     ioflo: Not Installed
                     PyZMQ: 14.0.1
                      RAET: Not Installed
                       ZMQ: 4.0.4
                      Mako: 0.9.1
                   Tornado: Not Installed
     Debian source package: 2015.5.3+ds-1trusty1
```

I'm trying to do this state:
```
docker-py:                                                                                                                                                                                          
    pkg.installed:                                                                                                                                                                                  
        - name: python-pip                                                                                                                                                                          
    pip.installed:                                                                                                                                                                                  
        - name: docker-py                                                                                                                                                                           
        - require:                                                                                                                                                                                  
            - pkg: python-pip   
```

Interesting to note - I *can* do 

```
salt 'thisminion' pip.installed docker-py==1.0.0
```

Just fine

**Comment by nitay** [comment]
On mine doing salt pip.installed didn't work. I found a workaround that works for me using get-pip.py to install pip instead of ubuntu: https://gist.github.com/nitay/a149724b5dffee95e8ac

Would be nice to have this solved but at least for the time being I'm not blocked on this.

**Comment by waynew** [comment]
@nitay that workaround works perfectly for me. Thanks!

**Comment by imchairmanm** [comment]
I'm running into the same issue here.

I'm not much of a programmer, but from what I can tell, the issue actually lies within `cmd.which_bin` which seems to be used to find `pip`.   If I run:

```
salt dev-web cmd.which_bin pip
```

All I get back is:

```
dev-web:
    None
```

This doesn't seem to be working correctly, since I do have `pip` installed:

```
salt dev-web cmd.run 'which pip'
```
```
dev-web:
    /usr/bin/pip
```

The same goes for many other executables.  For instance, I can type:

```
salt dev-web cmd.which_bin salt-minion
```

All I get back is:

```
dev-web:
    None
```

My systems are all on the same version:

```
salt-run manage.versions
```
```
Master:
    2015.8.0-365-g3ad48ca
Up to date:
    ----------
    dev-db:
        2015.8.0-365-g3ad48ca
    dev-web:
        2015.8.0-365-g3ad48ca
    salt:
        2015.8.0-365-g3ad48ca
    stage-db-01:
        2015.8.0-365-g3ad48ca
    stage-db-02:
        2015.8.0-365-g3ad48ca
    stage-lb-01:
        2015.8.0-365-g3ad48ca
    stage-lb-02:
        2015.8.0-365-g3ad48ca
    stage-web-01:
        2015.8.0-365-g3ad48ca
    stage-web-02:
        2015.8.0-365-g3ad48ca
```

And:

```
salt --versions-report
```
```
Salt Version:
           Salt: 2015.8.0-365-g3ad48ca
 
Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.0.1
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.4
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.3.0
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed
 
System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-57-generic
         system: Ubuntu 14.04 trusty
```

**Comment by remusao** [comment]
@imchairmanm The command `which_bin` only accept an iterable and not a single binary name, as stated in issue #21672 and the source code : https://github.com/saltstack/salt/blob/634264f443be075e0ad007fe8e40f3bfa8432c1b/salt/utils/__init__.py#L574. So it will always return an empty answer when given a single argument.

I'm also having the same problem with `pip` but I've not been able to work around given the `get-pip.py` script. Has any progress been made concerning this issue?

Thanks in advance,

**Comment by remusao** [comment]
It seems that after a fresh install of Ubuntu trusty 64 the problem is solved. My best guess is that maybe `pip.installed` requires another package than `python-pip`?

**Comment by tjeason** [comment]
I get this issue as well in my salt states if I don't include it before I install packages using pip: 

```
python-pip: 
  pkg.installed 
```

Salt installs pip version 1.0 on Ubuntu 12.04 LTs. But I need a newer version of pip. If I download pip manually or use Salt's `cmd.run` to automate the install, then pip is not ready by Salt to see if the package exists on the system.

**Comment by rallytime** [comment]
This is a regression in the state compiler for 2015.5.3 and it happens when you're running states that depend on the existence of an external package. Instead of the state gracefully falling and telling you something like `pip` is unavailable, it stack traces. This is fixed on 2015.5.6 and 2015.8.0 with #27681 and #25928, respectively.

Can anyone here test those versions and confirm this is fixed?

**Comment by rallytime** [comment]
Follow up to my last comment about pip.installed not working - I've submitted ##27996 to fix this problem. If anyone here could test these fixes and confirm they are  resolved for you, that'd be fantastic and much appreciated!

**Comment by rallytime** [comment]
Update: It appears that I mis-remembered when my back-porting PR went into the 2015.5 branch and it missed the 2015.5.6 release tag. However, the initial stacktrace reported by @nitay about the `KeyError` should be fixed at the HEAD of the 2015.5 branch along with the fix submitted in #27996, once it is merged in.

**Comment by rallytime** [comment]
Recap: There are a couple of things happening in this bug, so I am going to summarize them here with their fixes. The KeyError stacktrace was caused by the regression in 2015.5.3-6, mentioned above, and is fixed with #27681.

The problems seen where the pip installed on the machine imports in a shell but not with salt is due to an upstream bug in pip 1.5.4, which is what comes with Ubuntu 14.04 by default. More discussion about that can be found in #28036.

There can also be problems with using pip.installed on the 2015.5 branch where pip1 will fail if a package is already installed, which is fixed in #27996. 

Given that we haven't heard back on this issue for a bit, and that there are several fixes in place that will come out in 2015.5.7 and 2015.8.2, I am going to close this issue. If this pops up again, please leave a comment and we can certainly take another look. Thanks everyone!

**Comment by syedaali** [comment]
So I updated to 2015.8.1 to solve this problem, and I am now getting a different error:

salt-run manage.versions
'ZMQIOLoop' object has no attribute 'call_later'

**Comment by cachedout** [comment]
@syedaali How did you install 2015.8.1? This likely means you have a version of Tornado which is not recent enough.

**Comment by syedaali** [comment]
I was initially running 2015.5.3, which I had installed as follows:

$ sudo add-apt-repository ppa:saltstack/salt -y
$ sudo apt-get install python-software-properties -y
$ sudo apt-get install software-properties-common -y
$ sudo apt-get update -y
$ sudo apt-get install salt-master -y
$ sudo apt-get install salt-minion -y
$ sudo apt-get install salt-syndic -y

I then upgraded to 2015.8.1 as follows:

$ wget -O - https://repo.saltstack.com/apt/ubuntu/ubuntu14/latest/SALTSTACK-GPG-KEY.pub | sudo apt-key add -
$ echo "deb http://repo.saltstack.com/apt/ubuntu/ubuntu14/latest trusty main" >> /etc/apt/sources.list
$ sudo apt-get update
$ sudo apt-get install salt-master
$ sudo apt-get install salt-minion

**Comment by syedaali** [comment]
Here is what I currently have:

# dpkg --list | grep -i tornado
ii  python-tornado                   4.2.1-1                          amd64        scalable, non-blocking web server and tools

**Comment by syedaali** [comment]
More debug output:

# salt-run manage.versions -l all
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: ip-X-X-X-X.us-west-2.compute.internal
[TRACE   ] None of the required configuration sections, 'logstash_udp_handler' and 'logstash_zmq_handler', were found the in the configuration. Not loading the Logstash logging handlers module.
[TRACE   ] The required configuration section, 'fluent_handler', was not found the in the configuration. Not loading the fluent logging handlers module.
[DEBUG   ] Configuration file path: /etc/salt/master
[DEBUG   ] LazyLoaded manage.versions
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: ip-172-16-107-41.us-west-2.compute.internal
[DEBUG   ] Reading configuration from /etc/salt/master
[TRACE   ] Loading core.hwaddr_interfaces grain
[TRACE   ] Loading core.hostname grain
[TRACE   ] Loading core.get_master grain
[TRACE   ] Loading core.pythonversion grain
[TRACE   ] Loading core.path grain
[TRACE   ] Loading core.get_server_id grain
[TRACE   ] Loading core.ip6 grain
[TRACE   ] Loading core.ip4 grain
[TRACE   ] Loading core.saltversion grain
[TRACE   ] Loading core.saltpath grain
[TRACE   ] Loading core.pythonexecutable grain
[TRACE   ] Loading core.fqdn_ip4 grain
[TRACE   ] Loading core.fqdn_ip6 grain
[TRACE   ] Loading core.ip6_interfaces grain
[TRACE   ] Loading core.ip4_interfaces grain
[TRACE   ] Loading core.append_domain grain
[TRACE   ] Loading core.os_data grain
[TRACE   ] 'virt-what' could not be found in the following search path: ['/usr/local/sbin', '/usr/local/bin', '/usr/sbin', '/usr/bin', '/sbin', '/bin']
[DEBUG   ] Please install 'virt-what' to improve results of the 'virtual' grain.
[TRACE   ] 'systemd-detect-virt' could not be found in the following search path: ['/usr/local/sbin', '/usr/local/bin', '/usr/sbin', '/usr/bin', '/sbin', '/bin']
[TRACE   ] 'virt-what' could not be found in the following search path: ['/usr/local/sbin', '/usr/local/bin', '/usr/sbin', '/usr/bin', '/sbin', '/bin']
[TRACE   ] Loading core.zmqversion grain
[TRACE   ] Loading core.saltversioninfo grain
[TRACE   ] Loading core.pythonpath grain
[TRACE   ] Loading core.id_ grain
[TRACE   ] Loading core.locale_info grain
[TRACE   ] Loading core.get_machine_id grain
[TRACE   ] Loading core.ip_interfaces grain
[TRACE   ] Device ram0 does not report itself as an SSD
[TRACE   ] Device ram1 does not report itself as an SSD
[TRACE   ] Device ram2 does not report itself as an SSD
[TRACE   ] Device ram3 does not report itself as an SSD
[TRACE   ] Device ram4 does not report itself as an SSD
[TRACE   ] Device ram5 does not report itself as an SSD
[TRACE   ] Device ram6 does not report itself as an SSD
[TRACE   ] Device ram7 does not report itself as an SSD
[TRACE   ] Device ram8 does not report itself as an SSD
[TRACE   ] Device ram9 does not report itself as an SSD
[TRACE   ] Device xvda reports itself as an SSD
[TRACE   ] Device loop0 does not report itself as an SSD
[TRACE   ] Device loop1 does not report itself as an SSD
[TRACE   ] Device loop2 does not report itself as an SSD
[TRACE   ] Device loop3 does not report itself as an SSD
[TRACE   ] Device loop4 does not report itself as an SSD
[TRACE   ] Device loop5 does not report itself as an SSD
[TRACE   ] Device loop6 does not report itself as an SSD
[TRACE   ] Device loop7 does not report itself as an SSD
[TRACE   ] Device ram10 does not report itself as an SSD
[TRACE   ] Device ram11 does not report itself as an SSD
[TRACE   ] Device ram12 does not report itself as an SSD
[TRACE   ] Device ram13 does not report itself as an SSD
[TRACE   ] Device ram14 does not report itself as an SSD
[TRACE   ] Device ram15 does not report itself as an SSD
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Sending event - data = {'fun': 'runner.manage.versions', 'jid': '20151029191228926427', 'user': 'sudo_syed.ali', '_stamp': '2015-10-29T19:12:29.102570'}
[DEBUG   ] Error loading runners.nacl: libnacl import error, perhaps missing python libnacl package
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: ip-X-X-X-X.us-west-2.compute.internal
[DEBUG   ] Missing configuration file: /home/syed.ali/.saltrc
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'ip-X-X-X-X.us-west-2.compute.internal_master', 'tcp://127.0.0.1:4506', 'clear')
'ZMQIOLoop' object has no attribute 'call_later'
[DEBUG   ] LazyLoaded nested.output
[DEBUG   ] Error deleting fd from IOLoop
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/tornado/ioloop.py", line 554, in remove_handler
    self._impl.unregister(fd)
  File "/usr/lib/python2.7/dist-packages/zmq/eventloop/ioloop.py", line 125, in unregister
    return self._poller.unregister(fd)
  File "/usr/lib/python2.7/dist-packages/zmq/sugar/poll.py", line 80, in unregister
    idx = self._map.pop(socket)
KeyError: None
[TRACE   ] data = {}
[DEBUG   ] Sending event - data = {'jid': '20151029191228926427', 'return': {}, 'success': True, '_stamp': '2015-10-29T19:12:29.204458', 'user': 'sudo_syed.ali', 'fun': 'runner.manage.versions'}
[DEBUG   ] LazyLoaded local_cache.prep_jid
[INFO    ] Runner completed: 20151029191228926427
[DEBUG   ] Runner return: {}

**Comment by dragon788** [comment]
@syedaali According to rallytime on another issue the PPA lags behind their own repository. https://github.com/saltstack/salt/issues/30145#issuecomment-170360607
